trl/trainer/ppo_trainer.py

```diff
@@ -599,20 +639,30 @@ def batched_forward_pass(
                 if len(logprobs[j, start:end]) < 2:
                     raise ValueError("Responses are too short. Make sure they are at least 4 tokens long.")
 
-                all_values.append(v[j, start:end])
-                all_logprobs.append(logprobs[j, start:end])
-                all_ref_logprobs.append(ref_logprobs[j, start:end])
+                masks[j, :start] = 0
+                masks[j, end:] = 0
 
-        return all_logprobs, all_ref_logprobs, all_values
+            all_logits.append(logits)
+            all_values.append(values)
+            all_logprobs.append(logprobs)
+            all_masks.append(masks)
+
+        return (
+            torch.cat(all_logprobs),
+            torch.cat(all_logits)[:, :-1],
+            torch.cat(all_values)[:, :-1],
+            torch.cat(all_masks)[:, :-1],
+        )
 
```

Comment:
Since start/end is potentially different for each sample we wouldn't be able to stack them. The mask takes care of "slicing" by "masking" the tensor entries that are not used. the `[:-1]` is because we get predictions for one more token than we need. alternatively we could avoid this by removing the last `input_id` from the forward pass. No strong opinion :)

Reply:
I see this makes sense!